api/libproject/devstart.rb

```diff
@@ -413,6 +412,94 @@ def get_test_service_account_creds(*args)
   })
 end
 
+# Run commands with various gcloud setup/teardown: authorization and,
+# optionally, a CloudSQL proxy.
+class GcloudContext
+  def initialize(command_name, args, use_cloudsql_proxy = false)
+    @common = Common.new
+    @args = args
+    @parser = create_parser(command_name)
+    # Clients may access options to get default options (project etc)
+    # as well as their own custom options.
+    @opts = Options.new
+    @use_cloudsql_proxy = use_cloudsql_proxy
+  end
+
+  # Clients may override add_options and validate_options to add flags.
+  def add_options
+    add_default_options @parser, @opts
+  end
+
+  def validate_options
+    validate_default_options @parser, @opts
+  end
+
+  def run
+    add_options
+    @parser.parse @args
+    validate_options
+    ENV["GOOGLE_APPLICATION_CREDENTIALS"] = @opts.creds_file
+    begin
+      if @use_cloudsql_proxy
+        cloudsql_proxy_pid = run_cloud_sql_proxy(@opts.project, @opts.creds_file)
+        ENV["DB_PORT"] = "3307"
+      end
+
+      # FIXME add the wrapping for do_run_with_creds.
+      yield(self)
+    ensure
+      if @use_cloudsql_proxy
+        Process.kill("HUP", cloudsql_proxy_pid)
+      end
+    end
+  end
+end
+
+# Command-line parsing and main "run" implementation for set-authority.
+class SetAuthority < GcloudContext
+  # Adds command-line flags specific to set-authority.
+  def add_options
+    super
+    @parser.on(
+        "--email [EMAIL,...]",
+        "Comma-separated list of user accounts to change. Required."
+        ) do |email|
+      @opts.email = email
+    end
+    @parser.on(
+        "--add_authority [AUTHORITY,...]",
+        "Comma-separated list of user authorities to add for the users. " \
+        "One of added or removed authorities is required.") do |authority|
+      @opts.add_authority = authority
+    end
+    @parser.on(
+        "--rm_authority [AUTHORITY,...]",
+        "Comma-separated list of user authorities to remove from the users."
+        ) do |authority|
+      @opts.rm_authority = authority
+    end
+  end
+
+  def validate_options
+    super
+    if @opts.email == nil || (@opts.add_authority == nil && @opts.rm_authority == nil)
+      puts @parser.help
+      exit 1
+    end
+  end
+
+  def run
+    super do
+      Dir.chdir("tools") do
+        gradlew_path = File.join(Workbench::WORKBENCH_ROOT, "api", "gradlew")
+        @common.run_inline %W{
+            #{gradlew_path} --info setAuthority
+            -PappArgs="['#{@opts.email}','#{@opts.add_authority}','#{@opts.rm_authority}']"}
```

Comment:
Does this separate them with spaces? What happens if one of them is missing?

Reply:
(In general, I'm a little confused about how this is going to get parsed by the Java code...)

Reply:
The Java code would just get 3 positional args. Right now if one is missing you get an empty string, which I think will work out OK, but I haven't tried it yet.

I'm just looking for a way to front-load all the parsing/validation smarts, and keep everything in this particular function, and Gradle, and the Java, all as dumb as possible (meaning, requiring as little change as possible when you change the command-line flags).

Reply:
OK... I just wasn't sure how you'd distinguish between add_authority and rm_authority when there are only two non-empty args passed in. Is there going to be some prefix or something?

Reply:
An alternative that would be less ambiguous is:

<op> <emails> <authorities>

where <op> is "add" or "rm".

I don't think you really need to support adding and removing at the same time. At the moment we only have one authority anyway. :)